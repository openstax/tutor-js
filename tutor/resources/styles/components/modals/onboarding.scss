.onboarding {
  @include ost-hero-modal();
}

.onboarding.modal {
  @media screen and ( max-width: $mobile-collapse-breakpoint ){
    height: 100%;
  }
  .modal-dialog {
    width: 500px;
    margin: 0 auto;
    @media screen and ( max-width: $mobile-collapse-breakpoint ){
      width: 100%;
    }
  }
  &.show .modal-dialog {
    transform: translateY(130px);
    @media screen and ( max-width: $mobile-collapse-breakpoint ){
      transform: translateY(0) !important;
      height: 100%;
    }
  }
  .modal-content {
    border-radius: 0;
    @media screen and ( max-width: $mobile-collapse-breakpoint ){
      background-color: #f5f5f5;
      height: 100%;
    }
  }
  &.pay-now-or-later,
  &.free-trial-ended,
  &.make-payment {
    .modal-dialog {
      transform: translateY(80px);
    }
  }

  .onboarding-nag {
    background-color: $tutor-neutral-lighter;

    .heading {
      font-weight: 700;
      font-size: 18px;
      padding: 50px 50px 0 50px;
    }

    .body {
      font-weight: 500;
      font-size: 16px;
      line-height: 24px;
      padding: 50px;
      @media screen and ( max-width: $mobile-collapse-breakpoint ){
        padding: 50px 0;
      }
    }

    .footer {
      padding: 0 50px 50px 50px;
      .btn {
        width: 100%;
        margin-top: 0.5rem;
        & + .btn { margin-left: 0; }
      }
      &.got-it {
        .btn { width: inherit; }
      }
    }

    &.freshly-created-course {
      .btn-link { padding: 1rem 0; }
    }

    &.free-trial-ended,
    &.free-trial-activated
    {
      .footer {
        .btn { width: inherit; }
      }
    }

    &.pay-now-or-later,
    &.payments-disabled {
      padding: 50px;
      background-color: $tutor-neutral-bright;
      text-align: center;
      .btn.later {
        background-color: $tutor-neutral;
        color: $tutor-neutral-lightest;
      }
      p {
        font-size: 1.4rem;
        margin: 20px 0;

      }
      .or {
        overflow: hidden;
        text-align: center;
        margin: 20px 0;
        &:before,
        &:after {
          background-color: $tutor-neutral-light;
          content: "";
          display: inline-block;
          height: 1px;
          position: relative;
          vertical-align: middle;
          width: 50%;
        }
        &:before {
          right: 0.5em;
          margin-left: -50%;
        }
        &:after {
          left: 0.5em;
          margin-right: -50%;
        }
      }
    }
    &.payments-disabled {
      h3 { margin-bottom: 20px; }
    }
  }
}
