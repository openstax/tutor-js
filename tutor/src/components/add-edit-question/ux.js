import { action, observable, computed } from 'vendor';
import { filter, some, find } from 'lodash';

export default class AddEditQuestionUX {

  // local
  @observable showForm = false;

  // props
  @observable question;
  @observable book;
  // chapter/section ids
  @observable pageIds;

  //form
  @observable selectedChapter;
  @observable selectedChapterSection;

  constructor(props = {}) {
    this.book = props.book;
    // sections
    this.pageIds = props.pageIds;

    if(this.preSelectedChapters.length === 1) {
      this.selectedChapter = this.preSelectedChapters[0];
    }
  }

  @action.bound setShowForm(show) {
    this.showForm = show;
  }

  /**
   * Chapters that the user has selected
   */
  @computed get preSelectedChapters() {
    // return the chapters by checking the section pageIds
    return filter(this.book.chapters, bc => 
      some(bc.children, c => 
        some(this.pageIds, pi => pi === c.id)));
  }

  /**
   * Sections that the user has selected
   */
  @computed get preSelectedChapterSections() {
    if (!this.selectedChapter) {
      return null;
    }
    return filter(this.selectedChapter.children, scc => 
      some(this.pageIds, pi => pi === scc.id) && scc.isAssignable);
  }

  @action.bound setSelectedChapterByUUID(uuid) {
    this.selectedChapter = find(this.preSelectedChapters, psc => psc.uuid === uuid);
    if(this.preSelectedChapterSections.length === 1) {
      this.selectedChapterSection = this.preSelectedChapterSections[0];
    }
    else {
      this.selectedChapterSection = null;
    }
  }

  @action.bound setSelectedChapterSectionByUUID(uuid) {
    this.selectedChapterSection = find(this.preSelectedChapterSections, pscs => pscs.uuid === uuid);
  }
}
