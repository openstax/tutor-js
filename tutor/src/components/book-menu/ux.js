import { observable, action, observe, modelize } from 'shared/model'
import scrollIntoView from 'scroll-into-view';

export default class MenuUX {

    @observable currentPage;
    expanded = observable.map();

    constructor({ book, currentPage, wrapper }) {
        modelize(this);
        this.book = book;
        this.wrapper = wrapper;
        this.currentPage = currentPage;
        observe(this, 'currentPage', this.onPageChange, true);
    }

    expand(page, options = { wasSelected: true }) {
        this.eachParent(page, (node) => {
            const expandedOptions = this.expanded.get(node.pathId);
            if (!expandedOptions || !expandedOptions.wasSelected) {
                this.expanded.set(node.pathId, options);
            }
        });
    }

    collapse(page) {
        this.expanded.delete(page.pathId);
    }

    areSiblings(a, b) {
        return a.parent == b.parent;
    }

    eachParent(node, fn) {
        if (node) fn(node);
        while ((node = node.parent)) {
            fn(node);
        }
    }

    isExpanded(node) {
        return Boolean(this.expanded.get(node.pathId));
    }

    @action toggleExpansion(node, ev) {
        if (this.isExpanded(node)) {
            this.collapse(node);
        } else {
            this.expand(node);
        }
        ev.preventDefault();
    }

    wasAutoExpanded(page) {
        const options = this.expanded.get(page.pathId);
        return Boolean(options && options.wasSelected == false);
    }

    @action.bound onPageChange({ oldValue: oldPage, newValue: newPage }) {
        if (oldPage == newPage) { return; }

        // find chapters that were selected only because of the old page
        if (oldPage && !this.areSiblings(oldPage, newPage)) {
            this.eachParent(oldPage, (node) => {
                if (this.wasAutoExpanded(node)) {
                    this.collapse(node);
                }
            });
        }

        this.expand(newPage, { wasSelected: false });
        if (this.wrapper) {
            scrollIntoView(this.wrapper.querySelector(`[data-node-id="${newPage.pathId}"]`), {
                validTarget: (target) => {
                    return target == this.wrapper;
                },
            });
        }
    }

}
