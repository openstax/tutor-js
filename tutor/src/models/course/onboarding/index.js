import StudentCourse from './student-course';
import Preview from './preview';

export default function onboardingForCourse(course, context) {
    if (!course || (!course.currentRole.isStudentLike && !course.is_preview)) { return null; }

    const Klass = course.currentRole.isStudentLike ? StudentCourse : Preview;
    return new Klass(course, context);
}
