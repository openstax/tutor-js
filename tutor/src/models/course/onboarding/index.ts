import { StudentCourseOnboarding } from './student-course';
import { PreviewOnboarding } from './preview';
import { Course, TourContext } from '../../../models'

export default function onboardingForCourse(course: Course, context: TourContext) {
    if (!course || (!course.currentRole.isStudentLike && !course.is_preview)) { return null; }

    const Klass = course.currentRole.isStudentLike ? StudentCourseOnboarding : PreviewOnboarding;
    return new Klass(course, context);
}
