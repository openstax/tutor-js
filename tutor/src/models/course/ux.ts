import {
    computed, observable,
} from 'mobx';


export default class CourseUX {

    static get displayCourseCost() {
        return false;
    }

    static get formattedStudentCost() {
        return '$10';
    }

  @observable course;

  constructor(course) {
      modelize(this);
      this.course = course;
  }

  @computed get dataProps() {
      return {
          'data-title':       this.course.nameCleaned,
          'data-book-title':  this.course.bookName,
          'data-appearance':  this.course.appearance_code,
          'data-is-preview':  this.course.is_preview || false,
          'data-term':        this.course.termFull,
      };
  }

  @computed get courseType() {
      return this.course.is_concept_coach ? 'cc' : 'tutor';
  }

  @computed get courseId() {
      return this.course.isNew ? 'new' : this.course.id;
  }

  get formattedStudentCost() {
      return CourseUX.formattedStudentCost;
  }

}
