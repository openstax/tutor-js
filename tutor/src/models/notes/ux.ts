import { observable, action, computed } from 'mobx';
import { readonly } from 'core-decorators';

const ERROR_DISPLAY_TIMEOUT = 1000 * 4;

class StatusMessage {

  @observable display = false;
  @observable type;
  @observable message;
  @observable hideTimer;

  @computed get icon() {
      if (this.type === 'error') { return 'exclamation-triangle'; }
      return 'info-circle';
  }

  @action.bound hide() {
      this.display = false;
      this.hideTimer = null;
  }

  show({ message, type = 'error', autoHide = false }) {
      if (this.hideTimer) {
          clearTimeout(this.hideTimer);
          this.hideTimer = null;
      }

      Object.assign(this, {
          display: true, message, type,
      });

      if (autoHide) {
          this.hideTimer = setTimeout(
              this.hide, ERROR_DISPLAY_TIMEOUT
          );
      }
  }

}

class AnnotatorUX {

  @observable isSummaryVisible = false;

  @action.bound hideSummary() {
      this.isSummaryVisible = false;
  }

  @action.bound toggleSummary() {
      this.isSummaryVisible = !this.isSummaryVisible;
  }

  @readonly statusMessage = new StatusMessage();

}

const ux = new AnnotatorUX();

export default ux;
