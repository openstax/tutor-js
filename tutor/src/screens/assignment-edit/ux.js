import { React, observable, computed, action } from 'vendor';
import Router from '../../helpers/router';
import { runInAction } from 'mobx';
import ScrollTo from '../../helpers/scroll-to';
import {
  filter, isEmpty, compact, map, get, first, difference, flatMap, omit,
} from 'lodash';
import Exercises from '../../models/exercises';
import TaskPlan, { SELECTION_COUNTS } from '../../models/task-plans/teacher/plan';
import ReferenceBook from '../../models/reference-book';
import { StepUX, Step } from './step';
import { Actions } from './actions';
import Validations from './validations';
import moment from '../../helpers/moment-range';
import Time from '../../models/time';

const TEMPLATEABLE_TYPES = ['homework', 'reading'];

export default class AssignmentUX {

  @observable isShowingSectionSelection = false;
  @observable isShowingExerciseReview = false;
  @observable isShowingPeriodTaskings;
  @observable isShowingAddTemplate = false;
  @observable exercises;
  @observable isReady = false;
  @observable sourcePlanId;
  @observable form;
  @observable activeFilter = 'all';
  @observable templates;

  constructor(attrs = null) {
    if (attrs) { this.initialize(attrs); }
  }

  @action async initialize({
    id, type, plan, course, history, due_at,
    gradingTemplates = course.gradingTemplates,
    exercises = Exercises,
    windowImpl = window,
  }) {
    if ('clone' === type) {
      if (!course.pastTaskPlans.api.hasBeenFetched) {
        await course.pastTaskPlans.fetch();
      }
      this.sourcePlanId = id;
      this.plan = course.pastTaskPlans.get(id).createClone({ course });
    } else {
      if (plan) {
        this.plan = plan;
      } else {
        this.plan = new TaskPlan({ id, course, type });
        if (id && id != 'new') {
          const existing = course.teacherTaskPlans.get(id);
          if (existing) {
            this.plan.update( existing.serialize() );
          }
        }
      }
      if (type) {
        this.plan.type = type;
      }
    }
    this.course = course;
    // don't setup steps until course and plan is set
    this.steps = new StepUX(this);
    this.actions = new Actions(this);

    if (this.plan.isNew) {
      if (this.plan.isExternal || this.plan.isEvent) {
        const opens_at = moment(Time.now).add(1, 'day').startOf('day').add(1, 'minute').toISOString();
        this.periods.map((period) =>
          this.plan.findOrCreateTaskingForPeriod(period, { opens_at }),
        );
        if (due_at) {
          this.plan.tasking_plans.forEach(tp => {
            tp.initializeWithDueAt({ dueAt: due_at, defaultOpenTime: '00:01', defaultDueTime: '21:00' });
            tp.closes_at = moment(tp.due_at).add(1, 'minute').toISOString();
          });
        }
      } else {
        this.periods.map((period) =>
          this.plan.findOrCreateTaskingForPeriod(period),
        );
      }
    } else {
      await this.plan.ensureLoaded();
    }

    if (this.canSelectTemplates) {
      // once templates is loaded, select ones of the correct type
      await gradingTemplates.ensureLoaded();
      this.templates = gradingTemplates;
      this.plan.grading_template_id = this.plan.grading_template_id || get(this.gradingTemplates, '[0].id');
    }

    this.history = history;
    this.exercises = exercises;

    await this.exercises.ensureExercisesLoaded({ course, exercise_ids: this.plan.exerciseIds });

    if (this.plan.isReading) {
      await this.referenceBook.ensureLoaded();
    }
    this.windowImpl = windowImpl;

    this.isShowingPeriodTaskings = !(this.canSelectAllSections && this.plan.areTaskingDatesSame);

    this.scroller = new ScrollTo({ windowImpl });
    this.validations = new Validations(this);
    this.isReady = true;
  }

  @computed get gradingTemplates() {
    return(filter(this.templates.array, t => t.task_plan_type == this.plan.type));
  }

  @computed get gradingTemplate() {
    return (
      this.course.gradingTemplates.array.find(t => t.id == this.form.values.grading_template_id)
    );
  }

  @computed get canSelectAllSections() {
    return Boolean(
      this.plan.isEditable && isEmpty(difference(
        map(this.course.periods.active, 'id'),
        map(this.plan.tasking_plans, 'target_id'),
      ))
    );
  }

  @computed get canSelectTemplates() {
    return TEMPLATEABLE_TYPES.includes(this.plan.type);
  }

  @computed get canInputExternalUrl() {
    return this.plan.type === 'external';
  }

  @computed get periods() {
    return filter(this.course.periods.sorted, 'isActive');
  }

  @computed get isInitializing() {
    return !this.isReady || !this.plan || this.plan.api.isPendingInitialFetch;
  }

  get referenceBook() {
    if (this._referenceBook) {
      return this._referenceBook;
    }
    if (this.plan.ecosystem_id && this.plan.ecosystem_id != this.course.ecosystem_id) {
      this._referenceBook = new ReferenceBook({ id: this.plan.ecosystem_id });
    } else {
      this._referenceBook = this.course.referenceBook;
    }
    return this._referenceBook;
  }

  @action.bound renderStep(form) {
    this.form = form;
    return <Step ux={this} />;
  }

  @action.bound navigateToStep(index) {
    this.history.push(
      Router.makePathname('editAssignment', {
        courseId: this.course.id,
        type: this.plan.type,
        id: this.plan.id || 'new',
        step: index,
      })
    );
  }

  @action.bound onComplete() {
    this.history.push(`/course/${this.course.id}`);
  }

  @action.bound onCancel() {
    this.onComplete();
  }

  get formValues() {
    return this.plan.serialize();
  }

  @computed get isApiPending() {
    return this.plan.api.isPending;
  }

  @computed get canEdit() {
    return this.plan.canEdit;
  }

  @computed get selectedPageIds() {
    return this.plan.pageIds;
  }

  @computed get selectedChapterSections() {
    return filter(map(this.selectedPages, 'chapter_section'), 'isPresent');
  }

  @computed get selectedPages() {
    return this.selectedPageIds.map(pgId => this.referenceBook.pages.byId.get(pgId));
  }

  @action.bound onSectionIdsChange(ids) {
    this.plan.settings.page_ids = ids;
  }

  @action.bound togglePeriodTasking({ target: input }) {
    const period = this.plan.course.periods.find(p => p.id == input.dataset.periodId);
    if (!period) { return; }

    const tasking = this.plan.tasking_plans.forPeriod(period);

    if (input.checked && !tasking) {
      const firstTp = first(this.plan.tasking_plans);
      const opens_at = firstTp && firstTp.opens_at;
      const due_at = firstTp && firstTp.due_at;
      this.plan.findOrCreateTaskingForPeriod(period, { opens_at, due_at });
    } else if (!input.checked && tasking) {
      this.plan.tasking_plans.remove(tasking);
    }
  }

  @action.bound togglePeriodTaskingsEnabled(ev) {
    this.isShowingPeriodTaskings = ev.target.value == 'periods';
    if (this.isShowingPeriodTaskings) {
      return;
    }
    this.periods.map(period => this.plan.findOrCreateTaskingForPeriod(period));
  }

  @action.bound onExerciseToggle(event, exercise) {
    const ex = exercise.wrapper;
    ex.isSelected = !ex.isSelected;
    ex.isSelected ? this.plan.addExercise(ex) : this.plan.removeExercise(ex);
  }

  @computed get isFetchingExercises() {
    return this.exercises.isFetching({ pageIds: this.selectedPageIds });
  }

  @computed get isShowingExercises() {
    return this._stepIndex === 2;
  }

  @action async onExercisesReview() {
    await this.exercises.ensureExercisesLoaded({
      course: this.course,
      book: this.referenceBook,
      exercise_ids: this.plan.exerciseIds,
    });
  }

  @computed get selectedExercises() {
    if (isEmpty(this.exercises)) { return []; }
    return compact(map(this.plan.exerciseIds, exId => (
      this.exercises.get(exId)
    )));
  }

  @computed get numTutorSelections() {
    return get(this.plan, 'settings.exercises_count_dynamic', 0);
  }

  @action.bound increaseTutorSelection() {
    if (this.canIncreaseTutorExercises) {
      this.plan.settings.exercises_count_dynamic = this.numTutorSelections + 1;
    }
  }

  @action.bound decreaseTutorSelection() {
    if (this.canDecreaseTutorExercises) {
      this.plan.settings.exercises_count_dynamic = this.numTutorSelections - 1;
    }
  }

  @computed get canIncreaseTutorExercises() {
    return this.canEdit && this.numTutorSelections < SELECTION_COUNTS.max;
  }

  @computed get canDecreaseTutorExercises() {
    return this.canEdit && this.numTutorSelections > SELECTION_COUNTS.min;
  }

  @computed get numExerciseSteps() {
    return flatMap(this.plan.exercises, 'content.questions').length;
  }

  @computed get totalSelections() {
    return this.numExerciseSteps + this.numTutorSelections;
  }

  isExerciseSelected(ex) {
    return this.selectedExercises.includes(ex);
  }

  @computed get displayedExercises() {
    return this.exercises.where(ex => (
      this.isExerciseSelected(ex) || (
        ex.isHomework && ex.isAssignable &&
        ex.page && this.selectedPageIds.includes(ex.page.id)
      )
    ));
  }

  @action.bound async onSaveAsDraftClick() {
    await this.saveAsDraft();
    this.onComplete();
  }

  @action.bound async saveAsDraft() {
    this.plan.update(omit(this.form.values, 'tasking_plans'));
    this.plan.is_draft = true;
    await this.saveAndCopyPlan();
  }

  @action.bound async onPublishClick() {
    this.plan.update(omit(this.form.values, 'tasking_plans'));
    this.plan.is_draft = false;
    if (!this.plan.is_published) {
      this.plan.is_publish_requested = true;
    }
    await this.saveAndCopyPlan();
    this.onComplete();
  }

  @action async saveAndCopyPlan() {
    await this.plan.save();
    await runInAction(() => {
      const destPlan = this.course.teacherTaskPlans.withPlanId(this.plan.id);
      destPlan.update(this.plan.serialize());
    });
  }

  @action.bound onChangeFilter(value) {
    this.activeFilter = value;
  }

  @action.bound onShowAddTemplate() {
    this.isShowingAddTemplate = true;
  }

  @action.bound onHideAddTemplate() {
    this.isShowingAddTemplate = false;
  }

}
