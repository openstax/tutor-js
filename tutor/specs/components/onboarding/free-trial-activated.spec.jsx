import { C, Factory } from '../../helpers';
import FreeTrialActivated from '../../../src/components/onboarding/free-trial-activated';
import { StudentCourseOnboarding as Student } from '../../../src/components/onboarding/ux'

describe('Free trial ended modal', () => {
    let ux;
    beforeEach(() => {
        ux = new Student(Factory.course({ does_cost: true }))
    });

    it('renders and matches snapshot', () => {
        expect.snapshot(
            <C>
                <FreeTrialActivated
                    onDismiss={jest.fn()}
                    ux={ux} />
            </C>
        ).toMatchSnapshot();
    });

});
