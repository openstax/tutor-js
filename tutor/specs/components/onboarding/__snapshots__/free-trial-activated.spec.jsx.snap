// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Free trial ended modal renders and matches snapshot 1`] = `
<div
  className="openstax-debug-content"
>
  <div
    data-purpose="tour-conductor-wrapper"
  >
    <div>
      <div
        className="onboarding-nag free-trial-activated"
      >
        <div
          className="heading"
        >
          Your free trial is activated!
        </div>
        <div
          className="body"
        >
          <p>
            You will have access to 
            College Physics
             for 14 days.
          </p>
          <p>
            To continue accessing your course beyond your trial period, click the Pay Now button on your OpenStax Tutor Beta dashboard and enter your one-time 
            $10
             payment for the semester
          </p>
        </div>
        <div
          className="footer"
        >
          <button
            className="now btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Access your course
          </button>
        </div>
      </div>
    </div>
    <div
      className="visible-when-debugging"
    >
      <div
        className="tour-spy-info"
      >
        available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
      </div>
    </div>
  </div>
  <a
    aria-hidden="true"
    className="debug-toggle-link"
    href="#spy"
    onClick={[Function]}
    tabIndex={-1}
  >
    Ï€
  </a>
</div>
`;
